Cat Toy Shop 유효성 검사 / 회원관리 만들기

Flying-Fisherman opened the pull request:
강의에서 진행한 부분에 대해 수정하였습니다.
Web으로 출력될 회원 정보 부분에 대해 탐색하였습니다.

Flying-Fisherman committed:
Cat Toy Shop 유효성 검사 / 회원관리 만들기

강의에서 진행한 부분에 대해 수정하였습니다.
Web으로 출력될 회원 정보 부분에 대해 탐색하였습니다.

Flying-Fisherman committed:
회원관리 객체, DTO 완료, Service-DTO 작업 중

`AccountService`와 `Entity` 혹은 `DTO` 간의 데이터 처리 관련해서 로직 처리중입니다.
(AccountService 미완성)

johngrib commented on app/build.gradle:
```diff
@@ -51,6 +51,9 @@ dependencies {
     compileOnly 'org.projectlombok:lombok:1.18.16'
     annotationProcessor 'org.projectlombok:lombok:1.18.16'
 
+    // DozerMapper
+    implementation 'com.github.dozermapper:dozer-core:6.4.0'
```
코드 리뷰의 형식으로 DozerMapper에 대해 소개하는 글을 작성해 보세요.

Flying-Fisherman replied:
DozerMapper는 서로 다른 Java Bean 들을 재귀적으로 매핑(=연결) 해주는 Mapper입니다.
매핑되는 객체들은 동일한 속성 일 수도 있으나, 서로 다른 속성과 계층(Layer), 데이터 등의 복잡한 관계를 가지고 있으나
DozerMapper은 이를 간편하고 손 쉽게 매핑해주는 역할을 해줍니다.

사용 방법은 `@Annotations`, API를 기준으로 작성하겠습니다.


```
public class SourceBean {

    private Long id;

    private String name;

    @Mapping("binaryData")
    private String data;

    @Mapping("pk")
    public Long getId() {
        return this.id;
    }

    public String getName() {
        return this.name;
    }
}
```

`@Mapping` 어노테이션을 이용해 각 필드 or Getter에 대한 Mapping을 추가할 수 있습니다.

먼저 `id` 필드의 Getter에 올라간 `@Mapping`은 `pk`라는 이름의 속성을 찾게 됩니다. 
만약 `pk' 속성이 있다면, `id`와 `pk`는 양방향 Mapping이 이뤄지며, 자료형 변환이 가능한 경우, 두 가지 자료형의 데이터를 Mapping과 동시에 가지게 됩니다.

밑은 `pk`가 있는 class의 예시입니다
```
public class TargetBean {

    private String pk;

    private String name;

    private String binaryData;

    public void setPk(String pk) {
        this.pk = pk;
    }

    public void setName(String name) {
        this.name = name;
    }
}
```

즉 매핑된 `id`와 `pk`는 String과 Long의 자료형을 가지게 됩니다(자동적인 형변환이 되는 조건 하)

`SourceBean`의 `name`과 `TargetBean`의 `name`은 동일한 이름과 자료형으로 자동적으로 Mapping됩니다.

`SourceBean`의 `data`에 또한 어노테이션에 명시한 `binaryData`를 찾게되며, 이 또한 Mapping됩니다.


예시에 나와있는 `SourceBean`과 `TargetBean`은 세가지의 각 속성에 서로 매칭되어 Mapping이 되어 있으며, 이를 유동적으로 변환, 사용 할 수 있게 됩니다.




Script 출처 : https://dozermapper.github.io/gitbook/documentation/annotations.html [DozerMapper gitbook]

johngrib commented on app/src/main/java/com/codesoom/assignment/application/AccountService.java:
```diff
@@ -0,0 +1,48 @@
+
+@Service
+@Transactional
+public class AccountService {
```
`@Service`나 `@Controller`와 같은 주요 클래스에는 JavaDoc 주석을 작성해 주세요. 그리고 `public` 메소드에도 모두 JavaDoc 주석을 작성해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/AccountService.java:
```diff
@@ -0,0 +1,48 @@
+    private Object AccountData;
+
+    public AccountService(AccountRepository accRepo, Mapper dozerMapper) {
+        this.accRepo = accRepo;
```
`accRepo`와 같이 단어를 줄여쓰는 것을 지양해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Account.java:
```diff
@@ -0,0 +1,26 @@
+@Entity
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
lombok에서 `@NoArgsContructor`와 `@AllArgsConstructor`를 사용할 때 어떤 점들을 주의해야 하는지 조사해서 정리해 보세요.

Flying-Fisherman replied:
이건 내일 조사해서 정리하겠습니다 ㅠㅠ

Flying-Fisherman replied:
조사해보니 꽤나 다양한 주의점이 있었습니다!

`@NoArgsConstructor` 
해당 어노테이션은 파라미터가 없는 생성자를 만들어줍니다.

하지만 필드들이 `final`로 생성되어 있는 경우! 필드를 초기화 할 수 없어 생성자를 만들 수 없고, 에러가 발생합니다.
-> 이를 강제적으로 해결하기 위해서는 `@NoArgsConstructor(force=true)`로 옵션을 지정하여 강제로 초기화 하여 생성자를 만들 수 있습니다.

또한 `@NonNull`과 같이 필드에 제약조건이 설정되어 있는 경우, 생성자 내에 null-check 로직이 생성되지 않습니다.
즉, 처음 생성 후 초기화를 진행하면 null값에 의한 `NullPointerException`의 발생 가능성이 생깁니다.



`@AllArgsConstructor`
이 어노테이션은 클래스에 존재하는 모든 필드에 대한 생성자를 자동으로 만들어줍니다.
물론 `@NonNull` 같은 제약조건이 있을 경우 자동적으로 내부에서 null-check 로직이 생성됩니다.

발생 가능한 문제점으로는 lombok의 파라미터 순서와 필드 선언 순서 간의 문제가 발생합니다.

이 부분은 예시 코드를 이용하겠습니다.

```
@AllArgsConstructor
public static class Test() {
  private long param1;
  private long param2;
}
```

위와 같은 어노테이션과 생성자를 사용하면 자동적으로 생성되는 예시는 밑과 같게 됩니다.
```
Test test = new Test(1L, 2L);
```

만약 여기서 같은 타입의 필드들의 서순을 바꾸게 된다면?
```
@AllArgsConstructor
public static class Test() {
  private long param2;
  private long param1;
}
```

이 때 리팩토링이 동작하지 않고 lombok에서의 파라미터만 변환될 뿐이기 때문에
```
Test test = new Test(1L, 2L);
```

해당 생성자는 그대로 사용되게 되며
`param1 = 1L, param2 = 2L`이 되어야 하는 것이 단순한 필드 선언 순서의 변경으로
`param2 = 1L, param1 = 2L`이 되버립니다.

해당 문제는 `@AllArgsConstructor`뿐의 문제가 아니라 `Constructor`계열의 어노테이션의 공통적인 문제점으로 나타납니다.(`@Data`도 포함)

그래서 이를 해결하기 위해 `@Builder` 어노테이션이 있으며,
`@Builder`는 파라미터의 순서가 아닌 각 필드의 이름으로 값을 설정하여 리팩토링에도 유연, 필드 순서에 관계가 없어집니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/AccountData.java:
```diff
@@ -0,0 +1,29 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class AccountData {
+    private Long acc_uid;
```
`static final` 키워드를 사용한 상수를 쓰는 경우에만 이름에 `_`를 넣어주세요. 나머지 경우는 camelCase로 작성해주시고요.

johngrib commented on gradlew:
```diff
@@ -182,4 +182,4 @@ APP_ARGS=`save "$@"`
 eval set -- $DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS "\"-Dorg.gradle.appname=$APP_BASE_NAME\"" -classpath "\"$CLASSPATH\"" org.gradle.wrapper.GradleWrapperMain "$APP_ARGS"
 
-exec "$JAVACMD" "$@"
+exec "$JAVACMD" "$@"
```
불필요한 변경으로 보입니다.

Flying-Fisherman replied:
아 이 부분이 조금 애매한게... 마지막 개행문자가 있으면 실행시에 오류가 발생합니다.
`/usr/bin/env: ‘sh\r’: no such file or directory `

그래서 어쩔 수 없이 개행문자를 지우는데

아마 거기서 오는 변경 아닐까 싶습니다

johngrib replied:
그런가요? 그렇다면 일단 놔둡시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -4,6 +4,8 @@
 import com.codesoom.assignment.domain.Product;
 import com.codesoom.assignment.domain.ProductRepository;
 import com.codesoom.assignment.dto.ProductData;
+import com.github.dozermapper.core.DozerBeanMapperBuilder;
```
사용하지 않는 `import` 문입니다.

Flying-Fisherman replied:
여기까지 조언 주신 부분 체크했습니다!

Flying-Fisherman committed:
회원관리 객체, DTO, Service 수정 / Controller 생성

객체에 필요한 데이터 값 부터 수정하였습니다.
Controller 및 id관련 부분 수정 예정입니다.

미 테스트 입니다.

Flying-Fisherman committed:
JavaDoc 작성 / 잘못된 변수, 메서드 명 변경

johngrib commented on app/src/main/java/com/codesoom/assignment/application/AccountService.java:
```diff
@@ -0,0 +1,96 @@
+/**
+ * 회원정보에 대한 생성, 조회, 수정, 삭제 처리를 담당합니다.
+ */
+
```
```suggestion
```

johngrib commented on app/src/main/java/com/codesoom/assignment/application/AccountService.java:
```diff
@@ -0,0 +1,96 @@
+        this.mapper = dozerMapper;
+    }
+
+    /**
+     *
+     * @return 전체 회원정보입니다.
+     */
```
description을 적지 않으셨어요. 다음은 제가 작성해 본 것입니다. 나머지도 모두 작성해보세요.
```suggestion
    /**
     * 전체 회원정보 목록을 리턴합니다.
     * @return 회원정보 목록
     */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/application/AccountService.java:
```diff
@@ -0,0 +1,96 @@
+     * @param source 수정하고자 하는 회원정보 내용입니다.
+     * @return 수정된 회원정보입니다.
+     */
+    // TODO : Entity를 변경하기 위해서는 꼭 Entity 내에 Presentation Logic을 넣어서 사용해야 하는가?
```
이 문제를 해결하기 위한 다양한 패턴이 있습니다. 팩토리 패턴, 데코레이터 패턴, 빌더 패턴 등에 대해 조사해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/AccountController.java:
```diff
@@ -0,0 +1,100 @@
+ *
+ * 웹에서 회원정보에 대한 요청을 받아 처리 및 출력을 담당합니다.
+ *
+ * @author flying_fisherman
```
이 라인은 불필요합니다. 요즘 세상엔 git이 있어서 이런 게 별로 필요가 없죠.
```suggestion
```

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/AccountController.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.List;
+
+/**
+ *
+ * 웹에서 회원정보에 대한 요청을 받아 처리 및 출력을 담당합니다.
```
`회원 정보 조회, 변경에 대한 http 요청을 처리합니다.` 이렇게 하면 어떨까요? 물론 이건 제가 대충 생각해서 적은 거라 더 나은 문장을 작성하실 수도 있어요.

Flying-Fisherman replied:
Javadoc 작성하면서 궁금한 부분입니다만, 현재 프로젝트에서는 Controller과 Service 가 각각 하나씩 존재해서 동작 처리만 본다면 동일한 동작을 하는 것으로 볼 수 있을 것 같은데

이 부분에 대해서 JavaDoc을 작성할때 따로 분리를 해서 작성을 해 줘야 하는게 맞는걸까요?

johngrib replied:
물론이죠. 지금은 각각 하나씩 있다는 `우연` 때문에 그렇게 보이는 것입니다. 둘은 책임이 달라요. 컨트롤러는 http 의존을 담당하고 있고, 서비스는 비즈니스 로직을 담당하고 있죠. 그래서 컨트롤러에서는 http를 담당하고 있다는 것을 표현하는 것이 중요한데요, 누군가는 서비스에서 http response를 코딩하고 있다가 그것이 컨트롤러의 담당 범위라는 사실을 주석을 읽고 깨달아 자신이 실수하고 있다는 것을 파악하게 되겠죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/AccountController.java:
```diff
@@ -0,0 +1,100 @@
+ *
+ * @author flying_fisherman
+ */
+
```
```suggestion
```

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/AccountController.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    /**
+     * 전체 회원정보를 List화 하여 리턴합니다.
```
이렇게 하는 편이 더 낫습니다. `List화 하여`는 굳이 설명할 필요가 없어요. 이런 세부 구현에 대한 정보는 숨깁시다.
```suggestion
     * 전체 회원정보 List를 리턴합니다.
```
```suggestion
     * 전체 회원정보 목록을 리턴합니다.
```

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/AccountController.java:
```diff
@@ -0,0 +1,100 @@
+        return accountService.getAccounts();
+    }
+
+    /**
+     * id를 받아 해당 id의 회원정보를 리턴합니다.
+     *
+     * @param id 조회하려는 회원정보 ID입니다.
+     * @return 해당 ID의 회원정보입니다.
+     */
```
좋은 주석입니다. 그런데 `@param`에서 `id`에 대해 설명하고 있기 때문에 다음과 같이 축약할 수도 있어요.

```suggestion
    /**
     * 회원 정보를 리턴합니다.
     *
     * @param id 조회하려는 회원의 ID
     * @return 회원정보
     */
```

여기에서 생각해 볼 문제가 하나 더 있습니다. 만약에 존재하지 않는 id로 검색하려 한다면 이 메소드는 어떤 결과를 내놓아야 할까요?

Flying-Fisherman replied:
존재하지 않는 id로 검색하려 하는 경우, Service Layer의 `findAccount`를 통해 `AccountNotFoundException` 여부를 확인합니다.

JavaDoc 작성 시 거의 대부분의 메소드가 `findAccount`에 의존한다고 생각하여 `@throws` 태그는 해당 메소드에만 작성했습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Account.java:
```diff
@@ -0,0 +1,57 @@
+import javax.persistence.Id;
+
+/**
+ * 회원정보 Entity입니다.
```
짧고 단순하게 갑시다.
```suggestion
 * 회원정보.
```

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Account.java:
```diff
@@ -0,0 +1,57 @@
+
+    /**
+     * 회원의 이름입니다.
+     * 값은 Null이거나 공백일 수 없습니다.
```
강제할 수 없는 규칙을 주석으로 작성하는 것은 좋지 않습니다. 한 발 더 나아간다면 `Name`이라는 클래스를 만들어 엄격하게 정의할 수도 있겠죠.

Flying-Fisherman committed:
JavaDoc 수정 / 일부 ToDo 추가

Flying-Fisherman committed:
users API test 수정 / AccountServiceTest부터 생성 중

이전에 임창수 님이 올려주신 글을 읽고 오늘 test 동작 및 Script 일부를 수정했습니다.

이전 Test 관련한 내용을 복습하는 개념으로 로직을 짜 보고 있습니다.

Flying-Fisherman committed:
AccountServiceTest 추가 작성

johngrib commented on app/src/main/java/com/codesoom/assignment/application/AccountService.java:
```diff
@@ -0,0 +1,105 @@
+    /**
+     * 회원정보를 리턴합니다.
+     *
+     * @param id 조회하려는 회원정보 ID
```
`id`라는 문자열이 중복되는 것이 찝찝하다면 `식별자`라고 표현하는 방법도 있습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/AccountService.java:
```diff
@@ -0,0 +1,105 @@
+     * @return 작성된 회원정보
+     */
+    public Account createAccount(AccountData accountData) {
+//        return accountRepository.save(
+//                mapper.map(accountData, Account.class)
+//        );
```
이런 주석은 그냥 삭제해주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Account.java:
```diff
@@ -0,0 +1,54 @@
+import javax.persistence.Id;
+
+/**
+ * 회원정보
```
```suggestion
 * 회원정보.
```

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/AccountData.java:
```diff
@@ -0,0 +1,56 @@
+import com.github.dozermapper.core.DozerBeanMapperBuilder;
+import com.github.dozermapper.core.Mapper;
+import com.github.dozermapper.core.Mapping;
+import lombok.*;
```
`import *`의 사용을 지양해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/AccountData.java:
```diff
@@ -0,0 +1,56 @@
+
+    /**
+     * 회원의 이름입니다.
+     * Account의 name과 Mapping되어 있습니다.
+     * 값은 Null이거나 공백일 수 없습니다.
```
이 두 줄은 코드와 커플링이 생겨 있습니다. 삭제하는 것이 좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/AccountData.java:
```diff
@@ -0,0 +1,56 @@
+
+    /**
+     * 회원의 비밀번호입니다.
+     * Account의 password와 Mapping되어 있습니다.
+     * 값은 Null이거나 공백일 수 없습니다.
```
이 두 줄도 삭제해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/AccountData.java:
```diff
@@ -0,0 +1,56 @@
+
+    /**
+     * 회원의 이메일입니다.
+     * Account의 email와 Mapping되어 있습니다.
+     * 값은 Null이거나 공백일 수 없습니다.
```
삭제해 주세요.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceTest.java:
```diff
@@ -0,0 +1,129 @@
+        }
+
+        @Nested
+        @DisplayName("회원정보가 존재하지 않으면")
```
이렇게 하는 것이 더 낫겠습니다.
```suggestion
        @DisplayName("회원정보를 찾을 수 없다면")
```

왜냐하면, 준비 코드에서 `존재`를 준비하는 것이 아니라 `find`를 준비하고 있기 때문이에요.
```
            void setUp() {
                given(accountRepository.findAll())
                        .willReturn(List.of());
            }
```

Flying-Fisherman committed:
AccountServiceTest 작성 / JavaDoc 일부 수정

Flying-Fisherman commented:
이 전에 말씀해주신 '패턴'에 대해서도 조사를 해봤는데

생각보다 방식이나 방향성이 다양하고, 그 전에 제 머리가 이걸 원론적?으로 이해하는데 시간이 걸리기도 해서
정리 하는건 TIL이나 따로 포스팅을 해서 남겨야 할 것 같습니다.

전에 추천해주신 이펙티브 자바에도 같은 내용이 나오는데 아직은... 아직은 어렵습니다...

johngrib commented:
> 이 전에 말씀해주신 '패턴'에 대해서도 조사를 해봤는데
> 
> 생각보다 방식이나 방향성이 다양하고, 그 전에 제 머리가 이걸 원론적?으로 이해하는데 시간이 걸리기도 해서
> 정리 하는건 TIL이나 따로 포스팅을 해서 남겨야 할 것 같습니다.
> 
> 전에 추천해주신 이펙티브 자바에도 같은 내용이 나오는데 아직은... 아직은 어렵습니다...

이펙티브 자바 먼저 5번 읽어보세요 :-)
그 책은 반드시 읽어야 하는 책입니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceTest.java:
```diff
@@ -126,4 +140,143 @@ void it_throw_exception() {
+        }
+
+        @Nested
+        @DisplayName("허용되지 않은 문법을 전달받으면")
```
표현이 모호하게 느껴집니다. `허용되지 않는 문법`은 구체적으로 어떤 것인가요?

johngrib commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceTest.java:
```diff
@@ -126,4 +140,143 @@ void it_throw_exception() {
+    @DisplayName("createAccount 메서드")
+    class Describe_createAccount {
+        @Nested
+        @DisplayName("적합한 내용을 전달받으면")
```
`적합한 내용`은 무엇인가요? 어떤 타입인가요? 어떤 내용을 갖고 있나요?

johngrib commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceTest.java:
```diff
@@ -126,4 +140,143 @@ void it_throw_exception() {
+    class Describe_deleteAccount {
+
+        @Nested
+        @DisplayName("삭제하려는 등록된 회원정보의 id를 받으면")
+        class Context_with_existed_id {
+            @BeforeEach
+            void setUp() {
+                given(accountRepository.findById(REGISTERED_ID))
+                        .willReturn(Optional.of(account));
```
`setUp`의 코드와 비교해서 잘 읽어보세요. `findById`로 context를 설정했지요? 그런데 설명은 `삭제하려는 등록된 회원정보의 id를 받으면` 입니다. `deleteAccount`를 부른 이상 `삭제하려는`은 당연한 것일 거에요. 굳이 표기하지 않아도 괜찮죠. 한편 `등록된 회원정보`는 좀 더 컨텍스트 코드에 맞게 표현하는 방법이 있을 거에요.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceTest.java:
```diff
@@ -126,4 +140,143 @@ void it_throw_exception() {
+        }
+
+        @Nested
+        @DisplayName("존재하지 않는 회원정보의 id를 받으면")
+        class Context_with_not_existed_id {
+            @BeforeEach
+            void setUp() {
+                given(accountRepository.findById(NOT_REGISTERED_ID))
+                        .willThrow(AccountNotFoundException.class);
```
`findById` 부분을 읽고 잘 생각해 보세요. `존재하지 않는 회원정보의 id를 받으면`은 적절할까요?

다음과 같이 표현하면 어떨까요?
`찾을 수 없는 회원 id가 주어지면`

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 주간 회고를 작성해 공유해 주세요.
